src/tests/authentication.test.ts

```diff
@@ -0,0 +1,242 @@
+        const ipfsRepoPath = path.join(appDataPath);
+        const gqlPort = 15000
+        const hcPortAdmin = 15001
+        const hcPortApp = 15002
+        const ipfsSwarmPort = 15003
+
+        let agentCore: PerspectivismCore | null = null
+        let adminAd4mClient: Ad4mClient | null = null
+        let unAuthenticatedAppAd4mClient: Ad4mClient | null = null
+
+        beforeAll(async () => {
+            if (!fs.existsSync(appDataPath)) {
+                fs.mkdirSync(appDataPath, { recursive: true });
+            }
+
+            agentCore = await main.init({
+                appDataPath,
+                resourcePath: TEST_DIR,
+                networkBootstrapSeed: bootstrapSeedPath,
+                bootstrapFixtures: {
+                    languages: [],
+                    perspectives: [],
+                },
+                mocks: false,
+                gqlPort,
+                hcPortAdmin,
+                hcPortApp,
+                ipfsSwarmPort,
+                ipfsRepoPath,
+                hcUseBootstrap: false,
+                hcUseProxy: false,
+                hcUseLocalProxy: false,
+                hcUseMdns: true,
+                reqCredential: "123"
+            } as OuterConfig)
+
+            agentCore.initControllers()
+            await agentCore.initLanguages()
+
+            adminAd4mClient = new Ad4mClient(apolloClient(gqlPort, "123"))
+            await adminAd4mClient.agent.generate("passphrase")
+
+            unAuthenticatedAppAd4mClient = new Ad4mClient(apolloClient(gqlPort))
+        })
+
+        afterAll(async () => {
+            await agentCore!.exit();
+        })
+
+        it("unauthenticated user can not query agent status", async () => {
+            const call = async () => {
+                return await unAuthenticatedAppAd4mClient!.agent.status()
+            }
+
+            await expect(call())
+                .rejects
+                .toThrow("Capability is not matched");
+        })
+
+        it("unauthenticated user can request capability", async () => {
+            const call = async () => {
+                return await unAuthenticatedAppAd4mClient!.agent.requestCapability("demo-app", "demo-desc", "https://demo-link", '[{"with":{"domain":"agent","pointers":["*"]},"can":["QUERY"]}]')
+            }
+
+            await expect(call())
+                .resolves
+                .toMatch(/.+/)
+        })
+
+        it("admin user can permit capability", async () => {
+            let requestId = await unAuthenticatedAppAd4mClient!.agent.requestCapability("demo-app", "demo-desc", "https://demo-link", '[{"with":{"domain":"agent","pointers":["*"]},"can":["QUERY"]}]')
+            const call = async () => {
+                return await adminAd4mClient!.agent.permitCapability(`{"requestId":"${requestId}","auth":{"appName":"demo-app","appDesc":"demo-desc","appUrl":"demo-url","capabilities":[{"with":{"domain":"agent","pointers":["*"]},"can":["QUERY"]}]}}`)
```

Comment:
Hm, seeing the repetition of the capability request here in permitCapability makes it clear again for me, why I was arguing for the other solution (no permit function but finalising the request in generateJwt (which I think should be called finalise_request) by checking the random number - and instead having a decline function). 

I'm sorry I missed that when I approved the PR in ad4m with your set of functions yesterday, which means this solution is a bit more carved in stone.
But, as I said in our call, I really think this should be different.

Allowing for a cap request to return a capability that is different to what was requested only creates more complexity in the requesting UIs. So with this implementation, to avoid that complexity there, we need to make sure that any UI that calls `permitCapability` never changes the cap subject. Which is what AD4MIN definitely should do.

But also by not having a decline function, and not storing requests before they are permitted, UIs can't find out, if their request was declined.

I don't want to have another big turn-around on these PRs and I don't want to repeat myself again. And I don't like holding up the next steps of adding checks to all functions. So I want to request just one last change on this PR: can you please add comments (I guess best next to the definition of permitCapability) describing how this should be changed and why? Basically what I said in our calls and summarised above. Then I'm happy to merge this for now and let you continue on the rest of the capability story, and I will put up a PR myself in the not-too-distant future to change this around. 

Ok?

Reply:
Comments added, also link to the related issue here,
https://github.com/perspect3vism/ad4m-executor/issues/73